pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,270 @@
+					Debug("Skipping migration of egress rule due to failure in migration")
+				continue
+			}
+
+		case linux_defaults.RulePriorityEgressv2:
+			// Downgrade case. Unsupported for now.
+		}
+	}
+
+	return errors
+}
+
+// retrieveTableIDFromIfIndex computes the correct table ID based on the
+// ifindex provided. If table ID is comprised of the number associated with an
+// ENI device that corresponds to the ifindex.
+func retrieveTableIDFromIfIndex(ifindex int, getter interfaceDB) (int, error) {
+	link, err := netlink.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, err
+	}
+
+	ifaceNum, err := getter.GetInterfaceNumberByMAC(link.Attrs().HardwareAddr.String())
+	if err != nil {
+		return -1, err
+	}
+
+	return linux_defaults.RouteTableInterfacesOffset + ifaceNum, nil
+}
+
+// migrateOnUpgrade migrates the ENI routing table for the upgrade path. It
+// accepts the old table ID for the existing rules to migrate. `new` is the new
+// table ID that the rules should be referring to.
+func migrateOnUpgrade(old, new int) error {
+	if old == new {
+		// Nothing to do.
+		return nil
+	}
+
+	if err := copyRoutes(old, new); err != nil {
+		return err
+	}
+	if err := createNewRules(
+		linux_defaults.RulePriorityEgress,
+		old,
+		linux_defaults.RulePriorityEgressv2,
+		new,
+	); err != nil {
+		return err
+	}
+	if err := deleteOldRules(linux_defaults.RulePriorityEgress, old); err != nil {
+		return err
+	}
+	if err := deleteOldRoutes(old); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyRoutes(from, to int) error {
+	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+		Table: from,
+	}, netlink.RT_FILTER_TABLE)
+	if err != nil {
+		return err
+	}
+
+	// In ENI mode, we only expect two rules:
+	//  1) Link scoped route with a gateway IP
+	//  2) Default route via gateway IP
+	// We must upsert the link scoped route first which will make the kernel
+	// aware of the gateway IP. After that, then we will upsert the "default
+	// via gatewayIP" route. If we didn't do this, then the kernel would
+	// complain with "Error: Nexthop has invalid gateway." with an errno of
+	// ENETUNREACH.
+	for _, r := range routes {
+		if r.Scope == netlink.SCOPE_LINK {
+			route := r
+			route.Table = to
+			if err := netlink.RouteReplace(&route); err != nil {
+				return fmt.Errorf("unable to replace link scoped route under table ID: %w", err)
+			}
+
+			break
+		}
+	}
+
+	for _, r := range routes {
+		route := r
+		route.Table = to
+		if err := netlink.RouteReplace(&route); err != nil {
+			return fmt.Errorf("unable to replace route under table ID: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func createNewRules(fromPrio, oldTable, toPrio, newTable int) error {
+	rules, err := findRulesByPriority(fromPrio)
+	if err != nil {
+		return err
+	}
+
+	candidates := make([]netlink.Rule, 0, len(rules))
+	for _, r := range rules {
+		if r.Table == oldTable {
+			candidates = append(candidates, r)
+		}
+	}
+
+	for _, r := range candidates {
```

Comment:
Not sure why this 2nd loop is needed and why this is not just embedded into the above loop.

Reply:
I did this to guarantee that the link scoped route is inserted first. I didn't want to complicate the logic by putting it all into one loop. I figured the performance difference is probably negligible. Let me know if you think otherwise.

Reply:
^ Worth noting this is specifically to address the problem where we attempt to inject a route with next-hop of the gateway first, which will be rejected by the kernel as `-ENETUNREACH` because it doesn't know how to direct traffic hitting the route towards the getawy.

Reply:
Looking at this code again in context, this is the rules function not the routes function so I'm not sure I understand @christarazi 's response here. We don't need to create a list of candidates and then iterate them, we can just add rules in-place while iterating the list in the first place.